### PR TITLE
Iox #187 embedded roudi example

### PR DESCRIPTION
**Pull request by elfenpiff**

### CODE REVIEW

File: iceoryx_examples/README.md [hunk 1]
```diff
@@ -1,26 +1,28 @@
...
-
-    ./iceexample#          mkdir build && cd build
-    ./iceexample/build#    cmake .. -DCMAKE_INSTALL_PREFIX="`pwd`/../../../build/install/prefix"
-    ./iceexample/build#    cmake --build .
+| example                           | description |
+|:----------------------------------|:------------|
+|[icecrystal](./icecrystal/)        | Demostrates the usage of the iceoryx introspection client. |
+|[icedelivery](./icedelivery/)      | You are new to iceoryx then take a look at this example which demonstrates the basics of iceoryx by siding data from one process to another process. |
+|[iceperf](./iceperf/)              | A benchmark application which measures the latency of an IPC transmission between two applications. |
+|[singleprocess](./singleprocess/)  | Iceoryx can also be used for inter thread communication when you would like to run everything in a single process. |
```
Comment (mossmaurice) [inline iceoryx_examples/README.md:28]: Good stuff! Having two tables with the same content is not a good idea. Could you please remove the examples table in the main Readme.md and add a link to the new table? Thanks!

File: iceoryx_meta/CMakeLists.txt [hunk 2]
```diff
@@ -27,8 +27,27 @@ endif(introspection)
...
+message("          introspection.............: " ${introspection})
+message("          dds.......................: " ${dds})
+
+message("")
+message("       Build Properties")
+message("          project name..............: " ${CMAKE_PROJECT_NAME})
+message("          c compiler................: " ${CMAKE_C_COMPILER})
+message("          c++ compiler..............: " ${CMAKE_CXX_COMPILER})
+message("")
+
```
Comment (mossmaurice) [inline iceoryx_meta/CMakeLists.txt:53]: Nice!

File: iceoryx_posh/source/capro/service_description.cpp [hunk 3]
```diff
@@ -208,22 +208,6 @@ bool ServiceDescription::operator<(const ServiceDescription& rhs) const
...
-    m_serviceString = other.m_serviceString;
-    m_instanceString = other.m_instanceString;
-    m_eventString = other.m_eventString;
-    m_hasServiceOnlyDescription = other.m_hasServiceOnlyDescription;
-    m_classHash = other.m_classHash;
-    m_scope = other.m_scope;
-    m_interfaceSource = other.m_interfaceSource;
-
-    return *this;
-}
```
Comment (mossmaurice) [inline iceoryx_posh/source/capro/service_description.cpp:225]: Why this change?

File: iceoryx_examples/README.md [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
-built when using the script. If you want to build one example standalone add the path of the `build` directory when
-calling `cmake` after you have run the script:
-
-    ./iceexample#          mkdir build && cd build
-    ./iceexample/build#    cmake .. -DCMAKE_INSTALL_PREFIX="`pwd`/../../../build/install/prefix"
-    ./iceexample/build#    cmake --build .
+| example                           | description |
+|:----------------------------------|:------------|
+|[icecrystal](./icecrystal/)        | Demostrates the usage of the iceoryx introspection client. |
+|[icedelivery](./icedelivery/)      | You are new to iceoryx then take a look at this example which demonstrates the basics of iceoryx by siding data from one process to another process. |
```
Comment (MatthiasKillat) [inline iceoryx_examples/README.md:26]: Typo: siding->sending

File: iceoryx_posh/source/runtime/posh_runtime_single_process.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    auto currentFactory = PoshRuntime::s_runtimeFactory.target<PoshRuntime& (*)(const std::string&)>();
+    if (currentFactory != nullptr && *currentFactory == PoshRuntime::defaultRuntimeFactory)
+    {
+        PoshRuntime::s_runtimeFactory = [&](const std::string&) -> PoshRuntime& {
+            return *static_cast<PoshRuntime*>(this);
+        };
+    }
+    else
+    {
+        std::cerr << "PoshRuntimeSingleProcess can only created once per process and only if the default PoshRuntime "
```
Comment (dkroenke) [inline iceoryx_posh/source/runtime/posh_runtime_single_process.cpp:34]: Maybe Errorhandler here? just calling std::terminate is ugly.

File: iceoryx_posh/source/capro/service_description.cpp (hunk 3 above)
Comment (elfenpiff) [reply iceoryx_posh/source/capro/service_description.cpp:225]: Since the compiler generated version is identical.

File: iceoryx_examples/singleprocess/single_process.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+}
+
+int main()
+{
+    // set the log level to error to see the essence of the example
+    iox::log::LogManager::GetLogManager().SetDefaultLogLevel(iox::log::LogLevel::kError);
```
Comment (dkroenke) [inline iceoryx_examples/singleprocess/single_process.cpp:89]: Maybe we can pass the loglevel as argument to the main

File: iceoryx_posh/source/runtime/posh_runtime_single_process.cpp (hunk 5 above)
Comment (elfenpiff) [reply iceoryx_posh/source/runtime/posh_runtime_single_process.cpp:34]: You are right!

File: iceoryx_examples/singleprocess/single_process.cpp (hunk 6 above)
Comment (elfenpiff) [reply iceoryx_examples/singleprocess/single_process.cpp:89]: No for the example we only would like to see the interesting parts therefore I would leave it as it is.

File: iceoryx_examples/README.md [hunk 7]
```diff
@@ -1,26 +1,28 @@
...
-
-    ./iceexample#          mkdir build && cd build
-    ./iceexample/build#    cmake .. -DCMAKE_INSTALL_PREFIX="`pwd`/../../../build/install/prefix"
-    ./iceexample/build#    cmake --build .
+| example                           | description |
+|:----------------------------------|:------------|
+|[icecrystal](./icecrystal/)        | Demostrates the usage of the iceoryx introspection client. |
+|[icedelivery](./icedelivery/)      | You are new to iceoryx then take a look at this example which demonstrates the basics of iceoryx by sending data from one process to another process. |
+|[iceperf](./iceperf/)              | A benchmark application which measures the latency of an IPC transmission between two applications. |
+|[singleprocess](./singleprocess/)  | Iceoryx can also be used for inter thread communication when you would like to run everything in a single process. |
```
Comment (ithier) [inline iceoryx_examples/README.md:28]: Is "singleprocess" a better name than "embedded", which is the word usually used for such a thing (e.g. embedded web servers)?

File: iceoryx_examples/singleprocess/README.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    debug messages.
+
+    ```cpp
+    int main()
+    {
+        // set the log level to error to see the essence of the example
+        iox::log::LogManager::GetLogManager().SetDefaultLogLevel(iox::log::LogLevel::kError);
+    ```
+ 
+ 2. To start RouDi we have to create a configuration for him. We are choosing the
```
Comment (ithier) [inline iceoryx_examples/singleprocess/README.md:55]: him -> it

File: iceoryx_examples/singleprocess/README.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    int main()
+    {
+        // set the log level to error to see the essence of the example
+        iox::log::LogManager::GetLogManager().SetDefaultLogLevel(iox::log::LogLevel::kError);
+    ```
+ 
+ 2. To start RouDi we have to create a configuration for him. We are choosing the 
+    default config. Additionally, RouDi needs some other components like a memory 
+    management unit which handles how the memory is created in which the transmission
+    data is stored. The `IceOryxRouDiComponents` class is handling them for us
```
Comment (ithier) [inline iceoryx_examples/singleprocess/README.md:58]: is handling -> handles

File: iceoryx_examples/singleprocess/README.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    ```cpp
+    iox::RouDiConfig_t defaultRouDiConfig = iox::RouDiConfig_t().setDefaults();
+    iox::roudi::IceOryxRouDiComponents roudiComponents(defaultRouDiConfig);
+    ```
+
+ 3. We are starting RouDi and provide him with the required components. Furthermore, we
```
Comment (ithier) [inline iceoryx_examples/singleprocess/README.md:64]: him -> it
provide -> providing

File: iceoryx_examples/singleprocess/README.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    management unit which handles how the memory is created in which the transmission
+    data is stored. The `IceOryxRouDiComponents` class is handling them for us
+    ```cpp
+    iox::RouDiConfig_t defaultRouDiConfig = iox::RouDiConfig_t().setDefaults();
+    iox::roudi::IceOryxRouDiComponents roudiComponents(defaultRouDiConfig);
+    ```
+
+ 3. We are starting RouDi and provide him with the required components. Furthermore, we 
+    disable monitoring. The last bool parameter `false` states that RouDi does not 
+    terminate all registered processes when he goes out of scope. If we would set it
```
Comment (ithier) [inline iceoryx_examples/singleprocess/README.md:66]: he -> it

File: iceoryx_examples/singleprocess/README.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    std::this_thread::sleep_for(std::chrono::seconds(2));
+    keepRunning.store(false);
+
+    senderThread.join();
+    receiverThread.join();
+    ```
+
+### Implementation of Sender and Receiver
+Since there are no differences to inter process ports you can take a look at the
+[icedelivery example](../icedelivery) for a detailled documentation. We only provide
```
Comment (ithier) [inline iceoryx_examples/singleprocess/README.md:95]: detailled -> detailed

File: iceoryx_examples/singleprocess/README.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+#### Sender 
+We create a publisher port with the following service description 
+(Service = `Single`, Instance = `Process`, Event = `Demo`) and offer our service 
+to the world.
+```cpp
+iox::popo::Publisher publisher({"Single", "Process", "Demo"});
+publisher.offer();
+```
+After that we are sending
```
Comment (ithier) [inline iceoryx_examples/singleprocess/README.md:106]: are sending -> send

File: iceoryx_examples/singleprocess/README.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    sample->counter = counter++;
+    consoleOutput(std::string("Sending: " + std::to_string(sample->counter)));
+    publisher.sendChunk(sample);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+}
+```
+
+#### Receiver 
+Like with the sender we are creating a corresponding subscriber port with the
```
Comment (ithier) [inline iceoryx_examples/singleprocess/README.md:122]: are creating -> create
